subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileToolchainIntegrationTest.groovy

```diff
@@ -469,6 +469,55 @@ class JavaCompileToolchainIntegrationTest extends AbstractIntegrationSpec implem
         JavaVersion.current()   | "[deprecation] foo() in Foo has been deprecated"
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/23990")
+    @Requires(TestPrecondition.JDK11_OR_LATER)
+    def "can compile with a custom compiler executable"() {
+        def jdk = AvailableJavaHomes.getJdk(JavaVersion.current())
+
+        buildFile << """
+            plugins {
+                id("java")
+            }
+
+            configurations {
+                ecj {
+                    canBeConsumed = false
+                    canBeResolved = true
+                }
+            }
+
+            ${mavenCentralRepository()}
+
+            dependencies {
+                ecj("org.eclipse.jdt:ecj:3.32.0")
+            }
+
+            compileJava {
+                def customJavaLauncher = javaToolchains.launcherFor {
+                    languageVersion.set(JavaLanguageVersion.of(${jdk.javaVersion.majorVersion}))
+                }.get()
+
+                // ECJ does not support generating JNI headers
+                options.headerOutputDirectory.set(provider { null })
+                options.fork = true
+                options.forkOptions.executable = customJavaLauncher.executablePath.asFile.absolutePath
+
+                options.forkOptions.jvmArgumentProviders.add({
```

Comment:
When we do an example/test folks might copy from, let's add the ECJ configuration as a classpath input in this command line argument provider, so up-to-date checks will work. I think caching will still be problematic since the absolute path to the executable will be an input. Though at least up-to-date checks should work.

Reply:
Done